coreboot-toolchain: Introduce update script

felixsinger opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Simplify updating sources of that package.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

felixsinger committed:
coreboot-toolchain: Introduce script for generating sources file

Add and adjust update script from
https://git.petabyte.dev/petabyteboy/corenix.

The script is meant to be run from the package directory of the
coreboot-toolchain. The script generates a sources.nix files, which
contains all sources used for the coreboot toolchain and some other
tools. Thus, it needs to be stripped down to the necessary sources.

Signed-off-by: Felix Singer <felixsinger@posteo.net>

SuperSandro2000 commented on pkgs/development/tools/misc/coreboot-toolchain/update.sh:
```diff
@@ -0,0 +1,33 @@
+src="$(nix-build --no-out-link -A coreboot.src)"
+urls=$($src/util/crossgcc/buildgcc -u)
+
+echo '{ fetchurl }: [' > pkgs/coreboot/.files.nix.tmp
```
```suggestion
tmp=$(mktemp)
echo '{ fetchurl }: [' > $tmp
```

felixsinger replied:
Why not just this?

```
srcfile="sources.nix"
echo '{ fetchurl }: [' > ./$srcfile
```

SuperSandro2000 commented on pkgs/development/tools/misc/coreboot-toolchain/update.sh:
```diff
@@ -0,0 +1,33 @@
+	name="$(basename $url)"
+	hash="$(nix-prefetch-url "$url")"
+
+	cat << EOF >> pkgs/coreboot/.files.nix.tmp
+  {
+    name = "$name";
+    archive = fetchurl {
+      sha256 = "$hash";
+      url = "$url";
+    };
+  }
+EOF
+
+done
+
+echo ']' >> pkgs/coreboot/.files.nix.tmp
+mv pkgs/coreboot/.files.nix.tmp pkgs/coreboot/files.nix
```
```suggestion
	cat << EOF >> $tmp
  {
    name = "$name";
    archive = fetchurl {
      sha256 = "$hash";
      url = "$url";
    };
  }
EOF

done

echo ']' >> $tmp
mv $tmp pkgs/coreboot/files.nix
```

SuperSandro2000 commented on pkgs/development/tools/misc/coreboot-toolchain/update.sh:
```diff
@@ -0,0 +1,33 @@
+
+echo '{ fetchurl }: [' > pkgs/coreboot/.files.nix.tmp
+
+for url in $urls
+do
+	name="$(basename $url)"
+	hash="$(nix-prefetch-url "$url")"
```
```suggestion
for url in $urls; do
  name="$(basename $url)"
  hash="$(nix-prefetch-url "$url")"
```
Please use spaces instead of tabs for indentation.

SuperSandro2000 commented on pkgs/development/tools/misc/coreboot-toolchain/update-shell.nix:
```diff
@@ -0,0 +1,11 @@
+let
+  sources = import ../../nix/sources.nix;
+  pkgs = import sources.nixpkgs {};
+in
+
+with pkgs;
+
+mkShell {
+  buildInputs = [ nix git cacert getopt nix-update ];
+  NIX_PATH = "nixpkgs=${sources.nixpkgs}";
+}
```
Why do we need this niv file?

felixsinger replied:
Just for reference. Deleted.

felixsinger committed:
coreboot-toolchain: Use git repository as source

Signed-off-by: Felix Singer <felixsinger@posteo.net>

felixsinger committed:
coreboot-toolchain: Use sources.nix generated by update.sh

Use sources.nix for all source files (except coreboot) generated by
update.sh.

Signed-off-by: Felix Singer <felixsinger@posteo.net>

felixsinger commented:
@SuperSandro2000 Anything left to do?

maralorn reviewed (changes requested):
In general I like this change. Looks much cleaner than the code was before. I have one remark regarding `callPackage`.

When I run the script I get this failure:
```
[0.0 MiB DL]
error: unable to download 'https://downloads.sourceforge.net/sourceforge/expat/expat-2.2.9.tar.bz2': HTTP error 404
```
but on the other hand I get a lot more dependencies pulled in than are in your `sources.nix` what’s up with that?

When I run this script it outputs the `sources.nix` to the nixpkgs root, shouldn‘t it update the current file in place?

Also can you set the `passthru.updateScript` attribute?

maralorn commented:
As far as I can tell all my and Sandros concerns have been answered.

Thx, for maintaining the coreboot-toolchain!
